pkg/segment/query/processor/dataprocessor.go

```diff
@@ -69,7 +72,9 @@ func (dp *DataProcessor) SetStreams(streams []*CachedStream) {
 }
 
 func (dp *DataProcessor) Cleanup() {
+	dp.processorLock.Lock()
 	dp.processor.Cleanup()
+	dp.processorLock.Unlock()
```

Comment:
I think in addition to the lock, you'll also need to set a flag after Cleanup so we don't call Process after we've cleaned up. (Sorry I know I just told you to remove the bools; I didn't think about this case before)

Reply:
Yeah no worries. I thought about this case, but I thought that searcher would not send any data if the query is cancelled or deleted. 

But, I think the fetch would still continue calling processor.Process after it get the input from the streams. I have added this check.